presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimateMath.java

```diff
@@ -14,86 +14,196 @@
+
+        // everything will be filtered out after applying negation
+        if (outputRowCount == 0) {
+            return createZeroStats(superset);
+        }
+
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.builder();
+        result.setOutputRowCount(outputRowCount);
+
+        superset.getSymbolsWithKnownStatistics().forEach(symbol -> {
+            SymbolStatsEstimate supersetSymbolStats = superset.getSymbolStatistics(symbol);
+            SymbolStatsEstimate subsetSymbolStats = subset.getSymbolStatistics(symbol);
+
+            SymbolStatsEstimate.Builder newSymbolStats = SymbolStatsEstimate.builder();
+
+            // for simplicity keep the average row size the same as in the input
+            // in most cases the average row size doesn't change after applying filters
+            newSymbolStats.setAverageRowSize(supersetSymbolStats.getAverageRowSize());
+
+            // nullsCount
+            double supersetNullsCount = supersetSymbolStats.getNullsFraction() * supersetRowCount;
+            double subsetNullsCount = subsetSymbolStats.getNullsFraction() * subsetRowCount;
+            double newNullsCount = supersetNullsCount - subsetNullsCount;
+            verify(isNaN(newNullsCount) || newNullsCount >= 0, "newNullsCount must greater than or equal to zero: %s", newNullsCount);
+            newSymbolStats.setNullsFraction(min(newNullsCount, outputRowCount) / outputRowCount);
+
+            // distinctValuesCount
+            double supersetDistinctValues = supersetSymbolStats.getDistinctValuesCount();
+            double subsetDistinctValues = subsetSymbolStats.getDistinctValuesCount();
+            double newDistinctValuesCount;
+            if (isNaN(supersetDistinctValues) || isNaN(subsetDistinctValues)) {
+                newDistinctValuesCount = NaN;
+            }
+            else if (supersetDistinctValues == 0) {
+                newDistinctValuesCount = 0;
+            }
+            else if (subsetDistinctValues == 0) {
+                newDistinctValuesCount = supersetDistinctValues;
+            }
+            else {
+                double supersetNonNullsCount = supersetRowCount - supersetNullsCount;
+                double subsetNonNullsCount = subsetRowCount - subsetNullsCount;
+                double supersetValuesPerDistinctValue = supersetNonNullsCount / supersetDistinctValues;
+                double subsetValuesPerDistinctValue = subsetNonNullsCount / subsetDistinctValues;
+                if (supersetValuesPerDistinctValue <= subsetValuesPerDistinctValue) {
+                    newDistinctValuesCount = supersetDistinctValues - subsetDistinctValues;
+                }
+                else {
+                    newDistinctValuesCount = supersetDistinctValues;
+                }
+            }
+            verify(isNaN(newDistinctValuesCount) || newDistinctValuesCount >= 0, "newDistinctValuesCount must be greater or equal than zero: %s", newDistinctValuesCount);
```

Comment:
ditto, we should fail here, but we should add 0 check, see: https://github.com/prestodb/presto/pull/12060/commits/1d487a5e213128c6c572a2420ccd5a6820fd361f#diff-59fa6c2264eb72e0587e7e77ce155e59R90

Reply:
The same explanation as here: https://github.com/prestodb/presto/pull/11973#discussion_r241064493

Reply:
Please not forget about 0 check

Reply:
removed assertion